cli/src/pipeline.rs

```diff
@@ -99,21 +88,37 @@ pub struct FileImporter {
     errors: Vec<FileImportError>,
 }
 
+#[derive(Default)]
+pub struct SourcesCache {
+    cache: Vec<OwnedSource>,
+}
+
+impl SourcesCache {
+    pub fn extend(&mut self, sources: Vec<OwnedSource>) {
+        self.cache.extend(sources)
+    }
+
+    pub fn len(&self) -> usize {
+        self.cache.len()
+    }
+}
+
 impl FileImporter {
     /// Creates a new file importer that will import sources from the given
     /// root directory.
-    pub fn new(root: PathBuf) -> Self {
+    pub fn new(sources_shift: usize, root: PathBuf) -> Self {
```

Comment:
```suggestion
    pub fn new(root: PathBuf) -> Self {
```
A simple constructor would just start at `0`. To ensure `SourceId`s are different, just use the same importer and change the `root` `PathBuf` for instance.